### PR TITLE
 tests/gnrc_sock_dns: port to scapy

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This

- ports the existing tests to a shell based approach,
- that then also is used by a scapy version of the test, and
- adds failure tests to prevent a regression to before #10739.

### Testing procedure
Build and flash the application to a board of your choice (must support `ethos` or `netdev_tap`). Create a TAP interface and run

```sh
make test
```

The tests should succeed.

### Issues/PRs references
Adds tests for #10739 
~~Requires #10896~~ (merged) for tests to finish successfully.

### CODE REVIEW

File: tests/gnrc_sock_dns/README.md [hunk 1]
```diff
@@ -14,7 +14,6 @@ Start dnsmasq (in another console):
 
     $ sudo dnsmasq -d -2 -z -i tap0 -q --no-resolv \
         --dhcp-range=::1,constructor:tap0,ra-only \
-        --listen-address 2001:db8::1 \
```
Comment (miri64) [inline tests/gnrc_sock_dns/README.md]: I mentioned this in the commit that does this change: this bind causes (at least my version of `dnsmasq`) to send RAs from that address. However, sending RAs from a non-link-local address is invalid.

**Comment by miri64** [comment]
Rebased to current master to include fixes from #10896.

**Comment by aabadie** [comment]
> sudo make test

I have to say that I don't like this. The files generated by the build (if the application was not built already) will be owned by root and it will create a mess with file permissions in the riot directory. For me this is also a big problem with automated tests but also for developers without root access on their machine. The same issue is already in master with 2 other gnrc test applications.
The fact that it passes on Murdock is probably because the tests are run within a container environment, using a root user. This doesn't reflect the way a developer is supposed to use the build system to test something.
Running tests as root could cause many troubles, just imagine a line that just do `os.system('rm -rf /')`.

If there is a way to use scapy without the need to be root, that would be more useful and it should also be documented.

The purpose of tests is to be runnable by any developer, without strong knowledge of all the corner cases and setup details. It should also be easy to understand and help understand the code being tested.

**Comment by miri64** [comment]
The files generated by the build

**Comment by miri64** [comment]
If and only if I find a way to send `scapy`-cooked packets over UDP sockets, this test at least could be ported to UDP sockets, since it is testing DNS, an application layer protocol. I will try that. However, this is not possible with the other two tests (`gnrc_ipv6_ext` and `gnrc_rpl_srh`), since they test parts of the IPv6 implementation and are thus required to at least be able to send hand-cooked IPv6 datagrams.

**Comment by miri64** [comment]
> The purpose of tests is to be runnable by any developer, without strong knowledge of all the corner cases and setup details. It should also be easy to understand and help understand the code being tested.

Apart for developers without administrative access to the machine `scapy` allows for exactly that. As a bonus, the script even provides you a bold red warning if you try to run it without root permissions ;)

**Comment by miri64** [comment]
> `scapy` allows for exactly that.

To elaborate on that: the dependency-less alternative to `scapy` would be either the `struct` module or raw byte strings + sockets but I think those solutions are neither readable nor maintainable for the uninformed or informed developer ;)

**Comment by miri64** [comment]
> If and only if I find a way to send `scapy`-cooked packets over UDP sockets, this test at least could be ported to UDP sockets, since it is testing DNS, an application layer protocol. I will try that. However, this is not possible with the other two tests (`gnrc_ipv6_ext` and `gnrc_rpl_srh`), since they test parts of the IPv6 implementation and are thus required to at least be able to send hand-cooked IPv6 datagrams.

Worked. The tests are now even simpler than with pure `scapy` :-).

**Comment by miri64** [comment]
They still require a TAP interface though, so no way to run them with Murdock at the moment :(

**Comment by MrKevinWeiss** [comment]
Also rebase seems needed.

**Comment by miri64** [comment]
Rebased and adapted for master

**Comment by MrKevinWeiss** [comment]
Well since no response, I trust @miri64 that is testing the correct thing.

Please squash!

**Comment by miri64** [comment]
Squashed
